packages/@react-native-windows/codegen/Document.md

```diff
@@ -0,0 +1,386 @@
+```
+
+## Implementing a TurboModule in C++
+
+Checkout the following example:
+
+- Flow definition: [NativeDialogManagerWindows.js](https://github.com/microsoft/react-native-windows/blob/main/vnext/src/Libraries/Alert/NativeDialogManagerWindows.js)
+- Generated spec file: [NativeDialogManagerWindowsSpec.g.h](https://github.com/microsoft/react-native-windows/blob/main/vnext/codegen/NativeDialogManagerWindowsSpec.g.h)
+- Implementation: [AlertModule.h](https://github.com/microsoft/react-native-windows/blob/main/vnext/Microsoft.ReactNative/Modules/AlertModule.h)
+
+A TurboModule implemetation starts with:
+
+```c++
+#include <MyModuleSpec.g.h>
+#include <NativeModules.h>
+
+namespace MyNamespace
+{
+    REACT_MODULE(MyModule)
+    struct MyModule
+    {
+        using ModuleSpec = MyModuleSpec;
+    };
+}
+```
+
+You could call `winrt::Microsoft::ReactNative::MakeModuleProvider<MyNamespace::MyModule>`
+or `winrt::Microsoft::ReactNative::MakeTurboModuleProvider<MyNamespace::MyModule>`
+to create a `winrt::Microsoft::ReactNative::ReactModuleProvider` object and register it later.
+
+By calling `MakeModuleProvider<MyModule>`,
+if there is `using ModuleSpec = MyModuleSpec;` in `MyModule`,
+the compiler will compare `MyModule` with `MyModuleSpec` and see if all required members are defined.
+Failed to define any required member results in compiler errors telling you what member is needed.
+All acceptable member declarations in C++ for a missing member will be included in error messages,
+you could pick one and copy to your `MyModule`.
+
+`MakeTurboModuleProvider<MyModule>` does exactly what `MakeModuleProvider<MyModule>` does,
+but it requires `MyModule::ModuleSpec` to exist,
+unlike `MakeModuleProvider<MyModule>` where `MyModule::ModuleSpec` is optional.
+
+**Tips**: You could use `React::` if you think `winrt::Microsoft::ReactNative::` is too long after `#include <NativeModules.h>`.
+
+### REACT_MODULE
+
+The syntax of `REACT_MODULE` is `REACT_MODULE(moduleStruct, [optional]moduleName, [optional]eventEmitterName)`.
+
+- `moduleStruct`: the struct name that implements the TurboModule without namespace.
+- `moduleName`: the module name visible to JavaScript. The default value is `L"moduleStruct"`.
+- `eventEmitterName`: the default event emitter name. The default value is `L""`, which is `RCTDeviceEventEmitter`.
+
+### Members
+
+Any member in the struct implementing TurboModule functions should be attached with an attribute like `REACT_MODULE`.
+Here is the list of all available attributes:
+
+- `REACT_INIT`
+- `REACT_SYNC_METHOD`
+- `REACT_METHOD`
+- `REACT_GET_CONSTANTS`
+- `REACT_CONSTANT_PROVIDER`
+- `REACT_CONSTANT`
+- `REACT_EVENT`
+- `REACT_FUNCTION`
+
+### REACT_INIT
+
+`REACT_INIT` attaches to a function.
+Such function is called before a TurboModule is used in JavaScript,
+there is no deterministic timing about when this function will be called,
+and also no guarantee about in which thread this function will be called.
```

Comment:
It is not completely accurate. We call this method before we call any other members. The method is called after we initialize fields with the `REACT_EVENT` and `REACT_FUNCTION` attributes.
Currently this method is called from JS thread. 
In the past it used be called from UI thread.

Also, it is worth to mention that the main purpose of this method is to initialize the TurboModule and that the Initialize method receives ReactContext that provides access to RN instance data and services.

Reply:
I updated this piece of text. By the way, given that `IReactContext` has the ability to run code in specific thread, and `this method is called from JS thread` might be changed in the future (like you've mentioned that someone want to make a FastModel like thread model), so I will just keep the "no guarantee" here.